### PR TITLE
75 option to export csv from search

### PR DESCRIPTION
**Pull request by tom08**

### CODE REVIEW

File: exdb/templates/exdb/search.html [hunk 1]
```diff
@@ -14,6 +14,8 @@
     {{ block.super }}
 
     <div class="row">
+        <button id="export" class="button" data-url="{% url 'search_report' %}">Export as CSV</button>
```
Comment (enki211) [inline exdb/templates/exdb/search.html]: Trans tags

File: exdb/models.py [hunk 2]
```diff
@@ -136,6 +136,42 @@ def get_url(self, user):
...
         return reverse('edit', args=[self.pk])
 
+    def get_csv_row(self):
+        row = []
+        row.append(self.name)
+        row.append(self.get_status_display())
+        row.append(str(self.author))
+        planners = ''
+        for planner in self.planners.all():
+            planners += str(planner) + " "
```
Comment (shaneallgeier) [inline exdb/models.py]: This results in an extra space at the end of the string. Alternately, you could do:

```python
planners = ', '.join([str(p) for p in self.planners.all()])
```

File: exdb/models.py [hunk 3]
```diff
@@ -136,6 +136,42 @@ def get_url(self, user):
...
+        row.append(planners)
+        recognition = ''
+        for section in self.recognition.all():
+            recognition += section.name + " "
```
Comment (shaneallgeier) [inline exdb/models.py]: This results in an extra space at the end of the string. Alternately, you could do:

```python
recognition = ', '.join([s.name for s in self.recognition.all()])
```

File: exdb/models.py [hunk 4]
```diff
@@ -136,6 +136,42 @@ def get_url(self, user):
...
+        row.append(recognition)
+        row.append(self.start_datetime.strftime("%Y-%m-%d %H:%M"))
+        row.append(self.end_datetime.strftime("%Y-%m-%d %H:%M"))
+        row.append(self.type.name)
+        subtypes = ''
+        for subtype in self.subtypes.all():
+            subtypes += subtype.name + " "
```
Comment (shaneallgeier) [inline exdb/models.py]: This results in an extra space at the end of the string. Alternately, you could do:

```python
subtypes = ', '.join([st.name for st in self.subtypes.all()])
```

File: exdb/models.py [hunk 5]
```diff
@@ -136,6 +136,42 @@ def get_url(self, user):
...
+        row.append(subtypes)
+        row.append(self.description)
+        row.append(self.goals)
+        keywords = ''
+        for keyword in self.keywords.all():
+            keywords += keyword.name + " "
```
Comment (shaneallgeier) [inline exdb/models.py]: This results in an extra space at the end of the string. Alternately, you could do:

```python
keywords = ', '.join([k.name for k in self.keywords.all()])
```

File: exdb/models.py [hunk 6]
```diff
@@ -136,6 +136,42 @@ def get_url(self, user):
             return reverse('view_experience', args=[self.pk])
         return reverse('edit', args=[self.pk])
 
+    def get_csv_row(self):
```
Comment (shaneallgeier) [inline exdb/models.py]: You could change this to `convert_to_dict()` and return a dict for use with [csv.DictWriter](https://docs.python.org/3/library/csv.html#csv.DictWriter) in the view.

File: exdb/views.py [hunk 7]
```diff
@@ -383,3 +385,33 @@ def get_context_data(self, *args, **kwargs):
...
+        response['Content-Disposition'] = 'attachment; filename="experiences.csv"'
+
+        writer = csv.writer(response)
+        writer.writerow(['Experience Name', 'Status', 'Author', 'Planners', 'Recognition', 'Start Datetime',
+                         'End Datetime', 'Type', 'Subtypes', 'Description', 'Goals', 'Keywords', 'Audience',
+                         'Guest', 'Guest Office', 'Attendance', 'Created At', 'Next Approver', 'Funds', 'Conclusion',
+                         ])
+
+        for experience in experiences:
+            writer.writerow(experience.get_csv_row())
```
Comment (shaneallgeier) [inline exdb/views.py]: I recommend that you use [csv.DictWriter](https://docs.python.org/3/library/csv.html#csv.DictWriter) and convert `get_csv_row()` to return a dict instead.

File: exdb/views.py [hunk 8]
```diff
@@ -383,3 +385,36 @@ def get_context_data(self, *args, **kwargs):
         context = super(SearchExperienceResultsView, self).get_context_data(*args, **kwargs)
         context['search_query'] = self.request.GET.get('search', '')
         return context
+
+
+class SearchExperienceReport(CreateView):
```
Comment (shaneallgeier) [inline exdb/views.py]: This can simply inherit from `View` since you don't use anything that `CreateView` has to offer.

File: exdb/views.py [hunk 9]
```diff
@@ -383,3 +385,36 @@ def get_context_data(self, *args, **kwargs):
...
+        'next_approver', 'funds', 'conclusion',
+    ]
+
+    def get(self, *args, **kwargs):
+        if not self.request.GET.get('experiences'):
+            raise Http404
+        pks = json.loads(self.request.GET.get('experiences'))
+        if not pks:
+            raise Http404
+        experiences = Experience.objects.filter(pk__in=pks).prefetch_related(
```
Comment (shaneallgeier) [inline exdb/views.py]: We should still exclude the same drafts that we exclude from the regular search page.

**Comment by enki211** [comment]
Looks good

**Pull request merged**